nexus/types/src/deployment/diff_visitors/visit_blueprint_physical_disks_config.rs

```diff
@@ -0,0 +1,420 @@
+                    }
+                    map::Edit::Change { before, after, diff } => {
+                        v.visit_disk_change(ctx, Change::new(before, after));
+                        v.visit_disk_edit(ctx, diff);
+                    }
+                }
+            }
+            // Reset the context
+            ctx.disk_id = None;
+        }
+    }
+}
+
+pub fn visit_disk_edit<'e, V>(
+    v: &mut V,
+    ctx: &mut BpVisitorContext,
+    node: &EditedBlueprintPhysicalDiskConfig<'e>,
+) where
+    V: VisitBlueprintPhysicalDisksConfig<'e> + ?Sized,
+{
+    if let Edit::Change { before, after, .. } = &node.disposition {
+        v.visit_disk_disposition_change(ctx, Change::new(*before, *after));
+    }
+    if let Edit::Change { before, after, .. } = &node.identity {
+        v.visit_disk_identity_change(ctx, Change::new(*before, *after));
+    }
+    if let Edit::Change { before, after, .. } = &node.pool_id {
+        v.visit_disk_pool_id_change(ctx, Change::new(*before, *after));
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use diffus::Diffable;
+    use omicron_uuid_kinds::{PhysicalDiskUuid, ZpoolUuid};
+
+    struct TestVisitor<'a> {
+        before: &'a BlueprintPhysicalDisksConfig,
+        after: &'a BlueprintPhysicalDisksConfig,
+        total_inserts: usize,
+        total_removes: usize,
+        change_callbacks_fired: usize,
+    }
+
+    impl<'a> TestVisitor<'a> {
+        pub fn new(
+            before: &'a BlueprintPhysicalDisksConfig,
+            after: &'a BlueprintPhysicalDisksConfig,
+        ) -> Self {
+            TestVisitor {
+                before,
+                after,
+                total_inserts: 0,
+                total_removes: 0,
+                change_callbacks_fired: 0,
+            }
+        }
+    }
+
+    impl<'e> VisitBlueprintPhysicalDisksConfig<'e> for TestVisitor<'e> {
+        fn visit_generation_change(
+            &mut self,
+            ctx: &mut BpVisitorContext,
+            change: Change<'e, Generation>,
+        ) {
+            assert_eq!(change.before, &self.before.generation);
+            assert_eq!(change.after, &self.after.generation);
+            assert_ne!(self.before.generation, self.after.generation);
+
+            // We aren't operating on a specific disk
+            assert!(ctx.disk_id.is_none());
+
+            self.change_callbacks_fired += 1;
+        }
+
+        fn visit_disks_insert(
+            &mut self,
+            ctx: &mut BpVisitorContext,
+            node: &BlueprintPhysicalDiskConfig,
+        ) {
+            assert!(!self.before.disks.contains_key(&node.id));
+            assert!(self.after.disks.contains_key(&node.id));
+
+            // The inserted node is the same as what's in `after`
+            assert_eq!(node, self.after.disks.get(&node.id).unwrap());
+
+            // The key for the current disk id was filled in
+            assert_eq!(ctx.disk_id, Some(node.id));
```

Comment:
Idle thoughts that you may have already considered and discarded for good reason - it seems like the fields of `ctx` are expected to always be `Some(_)` or always be `None` depending on which visitor method we're in. Does that imply that different visitor methods could take different kinds of context objects to remove the optionality?

Reply:
This was the first thing I thought of when discussing with @sunshowers last week. However, the goal really is to be optional across everything, because we are going to be composing the visitors and this allows us to get context inside the callbacks from different visitors higher up the tree.